proxy/src/opentsdb/types.rs

```diff
@@ -0,0 +1,226 @@
+
+use crate::error::{ErrNoCause, ErrWithCause, Result};
+
+const OPENTSDB_DEFAULT_FIELD: &str = "value";
+
+#[derive(Debug)]
+pub struct PutRequest {
+    pub points: Bytes,
+
+    pub summary: Option<String>,
+    pub details: Option<String>,
+    pub sync: Option<String>,
+    pub sync_timeout: i32,
+}
+
+impl PutRequest {
+    pub fn new(points: Bytes, params: PutParams) -> Self {
+        PutRequest {
+            points,
+            summary: params.summary,
+            details: params.details,
+            sync: params.sync,
+            sync_timeout: params.sync_timeout,
+        }
+    }
+}
+
+pub type PutResponse = ();
+
+/// Query string parameters for put api
+///
+/// It's derived from query string parameters of put described in
+/// doc of OpenTSDB 2.4:
+///     http://opentsdb.net/docs/build/html/api_http/put.html#requests
+///
+/// NOTE:
+///     - all the params is unimplemented.
+#[derive(Debug, Default, Deserialize)]
+#[serde(default)]
+pub struct PutParams {
+    pub summary: Option<String>,
+    pub details: Option<String>,
+    pub sync: Option<String>,
+    pub sync_timeout: i32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Point {
+    pub metric: String,
+    pub timestamp: i64,
+    pub value: Value,
+    pub tags: HashMap<String, String>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub enum Value {
+    IntegerValue(i64),
+    F64Value(f64),
+    StringValue(String),
+}
+
+impl Value {
+    fn try_to_f64(&self) -> std::result::Result<f64, ParseFloatError> {
+        match self {
+            Value::IntegerValue(v) => Ok(*v as f64),
+            Value::F64Value(v) => Ok(*v),
+            Value::StringValue(v) => v.parse(),
+        }
+    }
+}
+
+impl Display for Value {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Value::IntegerValue(v) => Display::fmt(v, f),
+            Value::F64Value(v) => Display::fmt(v, f),
+            Value::StringValue(v) => Display::fmt(v, f),
+        }
+    }
+}
+
+pub(crate) fn convert_put_request(req: PutRequest) -> Result<Vec<WriteTableRequest>> {
+    let points = {
+        // multi points represent as json array
+        let parse_array = from_slice::<Vec<Point>>(&req.points);
+        match parse_array {
+            Ok(points) => Ok(points),
+            Err(_e) => {
+                // single point represent as json object
+                let parse_object = from_slice::<Point>(&req.points);
+                match parse_object {
+                    Ok(point) => Ok(vec![point]),
+                    Err(e) => Err(e),
+                }
+            }
+        }
+    };
+    let points = points.box_err().with_context(|| ErrWithCause {
+        code: StatusCode::BAD_REQUEST,
+        msg: "Json parse error".to_string(),
+    })?;
+    validate(&points)?;
+
+    let mut points_per_metric = HashMap::new();
+    for point in points.into_iter() {
+        points_per_metric
+            .entry(point.metric.clone())
```

Comment:
Only use metric name as hash key may not enough, such as
```
A{host="xx"}
A{host="xx",region="yy"}
```
This should be put into different WriteSeriesEntry, you can refer how prom write convert works:
- https://github.com/CeresDB/ceresdb/blob/0a26ad2098a1d4b3e742139808c75bc0dc302a61/proxy/src/http/prom.rs#L403

Reply:
I think they can be put into the same WriteSeriesEntry which contains all the presented tags.
If a row didn't contains a tag, it will take a null value.
```
A{host="xx", region=null}
A{host="xx",region="yy"}
```

Reply:
Yes, they can be put together like you said, but did you tested this case? 

I think you can add unit tests for this function `convert_put_request` to ensure everything works as expected.

Reply:
The test plan as said above has tested this case. I will add more tests in other case.